### PR TITLE
add component module concept

### PR DESCRIPTION
**Pull request by sns-seb**
this PR introduces the concept of module as a group of components
this will allow reducing length of the ComponentLevel classes and in the near future implement lazy start of components and/or description of implicit component dependencies
this PR also includes a few module to demo usage and as a proof-of-concept

*this PR is dependent on PR #307*

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/platform/components/ComponentLevel4.java [hunk 1]
```diff
@@ -324,7 +318,7 @@ public ComponentLevel4(ComponentLevel parent, List<Object> level4AddedComponents
   }
 
   @Override
-  public ComponentLevel configure() {
+  protected void configureLevel() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/components/ComponentLevel4.java]: :new: MAJORThis method has 389 lines, which is greater than the 100 lines authorized. Split it into smaller methods.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)

File: server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.sonar.core.platform.ComponentContainer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public abstract class Module {
+  private ComponentContainer container;
+
+//  public void start() {
+//    // forces instantiation and start of implicit dependencies
+//    for (Class<?> aClass : implicitDependencies()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+//      container.getPicoContainer().getComponent(aClass);
+//    }
+//  }
+//
+//  public Collection<Class<?>> implicitDependencies() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.sonar.core.platform.ComponentContainer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public abstract class Module {
+  private ComponentContainer container;
+
+//  public void start() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 4 major

Details:
* [This method has 389 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonarqube/blob/dd5192c304ff20e4656e59c99094415fbc43f4bb/server/sonar-server/src/main/java/org/sonar/server/platform/components/ComponentLevel4.java#L321)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonarqube/blob/dd5192c304ff20e4656e59c99094415fbc43f4bb/server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java#L33)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonarqube/blob/dd5192c304ff20e4656e59c99094415fbc43f4bb/server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java#L38)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonarqube/blob/dd5192c304ff20e4656e59c99094415fbc43f4bb/server/sonar-server/src/main/java/org/sonar/server/platform/components/Module.java#L31)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by sns-seb** [comment]
I'm wondering whether `ComponentLevel` should be moved to sonar-core to benefit the batch

File: server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.batch;
+
+import org.sonar.core.component.Module;
+import org.sonar.server.computation.ws.SubmitReportAction;
+
+public class BatchWsModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.search;
+
+import org.sonar.core.component.Module;
+import org.sonar.server.es.EsClient;
+
+public class EsSearchModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 7]
```diff
@@ -306,7 +299,7 @@ public PlatformLevel4(PlatformLevel parent, List<Object> level4AddedComponents)
   }
 
   @Override
-  public PlatformLevel configure() {
+  protected void configureLevel() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: :new: MAJORThis method has 387 lines, which is greater than the 100 lines authorized. Split it into smaller methods.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)

File: server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java [hunk 8]
```diff
@@ -73,78 +73,77 @@
 import org.sonar.server.db.migrations.v52.FeedProjectLinksComponentUuid;
 import org.sonar.server.db.migrations.v52.MoveProjectProfileAssociation;
 
-import java.util.List;
+public class MigrationStepModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 4 major

Details:
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/88fa8664055fd03279506f08f786a291ec15e0fb/server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java#L27)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/88fa8664055fd03279506f08f786a291ec15e0fb/server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java#L78)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* [This method has 387 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonarqube/blob/88fa8664055fd03279506f08f786a291ec15e0fb/server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java#L302)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/88fa8664055fd03279506f08f786a291ec15e0fb/server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java#L27)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by sns-seb** [comment]
updated the PR
* rebased on master since #307 has been merged
* moved Module to core (which trigger FP issues "Protected method 'configureModule(...)' is never used")

**Pull request closed without merging**